0005-R-pluggable-components-injector.md

```diff
@@ -0,0 +1,298 @@
+| Persona           | Cluster operator/End user                                                                | End user                                                                                                            |
+| Scope             | Does not need to be scoped                                                               | If not scoped, all applications should have deployed the pluggable component, otherwise runtime errors might happen |
+
+</div></b>
+
+#### Component spec annotations
+
+The component spec is still the entry point for all component types being pluggable or not, given that the pluggable components are a subset of all users declared components, even more, the pluggable components can be inferred from the declared components, we can actually leverage that property to extend our component spec, by adding custom annotations to allow Dapr to inject the component container at the time the Injector is also injecting the Dapr sidecar container.
+
+Example:
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: my-component
+  annotations:
+    dapr.io/component-container-image: "component:v1.0.0"
+spec:
+  type: state.my-componentƒ
+  version: v1
+  metadata: []
+```
+
+Optionally you can mount volumes and add env variables into the containers by using the `dapr.io/component-container-volume-mounts(-rw)` and `dapr.io/component-container-env` annotations.
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: my-component
+  annotations:
+    dapr.io/component-container-image: "component:v1.0.0"
+    dapr.io/component-container-volume-mounts: "volume-name:/volume-path,volume-name-2:/volume-path-2" # read-only, "$VOLUME_NAME:$VOLUME_PATH,$VOLUME_NAME_2:$VOLUME_PATH2"
+    dapr.io/component-container-volume-mounts-rw: "volume-name-rw:/volume-path-rw,volume-name-2-rw:/volume-path-2-rw" # read-write "$VOLUME_NAME:$VOLUME_PATH,$VOLUME_NAME_2:$VOLUME_PATH2"
+    dapr.io/component-container-env: "env-var=env-var-value,env-var-2=env-var-value-2" #optional "$ENV_NAME=$ENV_VALUE,$ENV_NAME_2=$ENV_VALUE_2"
+spec:
+  type: state.my-component
+  version: v1
+  metadata: []
+```
+
+By default the injector creates undeclared volumes as `emptyDir` volumes, if you want a different volume type you should declare it by yourself in your pods.
+
+#### Pod annotations
+
+In order to allow users to turn off the component injector for their pod, a new annotation will be available, similar to the one that we have for enabling dapr: `dapr.io/inject-pluggable-components:"true"`. Let's rewrite the previous examples using the injected pluggable components feature, it would be something like:
+
+The apps deployments/pods:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+name: app
+labels:
+  app: app
+spec:
+replicas: 1
+selector:
+  matchLabels:
+    app: app
+template:
+  metadata:
+    labels:
+      app: app
+    annotations:
+      dapr.io/inject-pluggable-components: "true"
```

Comment:
I don't see how this solves the scoping problem. What happens if you have 2 pluggable component and two apps - each app needing a different pluggable component? This annotation would inject both pluggable components to a deployment.

Reply:
> This annotation would inject both pluggable components to a deployment.
Yes, this annotation hasn't the purpose of solving scope, it is basically used to enabling/disabling the pluggable components injector, similar from what we have for the sidecar injector.

The "scope" problem that is solved by this proposal is the similarity with how the built-in components work today. You don't have to worry about scope them because it just works. Differently from what pluggable components offers, which kinda obligates the user to properly scope pluggable components due to the container's dependency on their pods.